crowbar_framework/app/models/api/node.rb

```diff
@@ -50,6 +50,32 @@ def repocheck(options = {})
       end
     end
 
+    def already_upgraded?(node)
+      current_platform = "#{node[:platform]}-#{node[:platform_version]}"
+      return false if current_platform != node[:target_platform]
+
+      Rails.logger.info("node #{node.name} is already upgraded")
+      true
+    end
+
+    def pre_upgrade(node)
+      # Migrate out the l3 agents and shut down pacemaker
+      # FIXME: should we make thise code part of crowbar-upgrade-os.sh?
+      node.ssh_cmd("/usr/sbin/crowbar-pre-upgrade.sh")
+    end
+
+    def upgrade(node)
+      # Upgrade one node
+      # TODO: signal to the UI that we started upgrade of the node
+      node.ssh_cmd("/usr/sbin/crowbar-upgrade-os.sh")
+    end
+
+    def post_upgrade(node)
+      # Join the cluster: start pacemaker and run selected recipes
+      # Migrate l3 agents to this node
```

Comment:
Migrating L3 agents back here is not necessary - that will happen anyway if necessary when the next node does its pre-upgrade step. When evacuating L3 agents from a node, we don't want to specify a target because we want to rely on the built-in intelligence of neutron to pick the best destination for the migration, in order to keep things nicely balanced.